src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java

```diff
@@ -0,0 +1,617 @@
+      desiredRotation = 0;
+    }
+
+    drive(x, y, desiredRotation, fieldRelative);
+  }
+
+  /**
+   * Rotates to a specific angle from the D-pad buttons
+   * 
+   * @param povAngleDeg Angle of D-pad control
+   */
+  public int convertCardinalDirections(int povAngleDeg) {
+    // TODO: figure out what angles the robot should turn to
+    if (povAngleDeg == 270) {
+      povAngleDeg += 90;
+    } else if (povAngleDeg == 90) {
+      povAngleDeg -= 90;
+    }
+    // targetHeadingDegrees is counterclockwise so need to flip povAngle
+    povAngleDeg = 360 - povAngleDeg;
+    return povAngleDeg;
+  }
+
+  /**
+   * Determines whether to rotate according to input or to run the keep heading
+   * code, by checking if
+   * the (already deadbanded) rotation input is equal to 0.
+   *
+   * @param x             Desired speed of the robot in the x direction (forward),
+   *                      [-1,1].
+   * @param y             Desired speed of the robot in the y direction
+   *                      (sideways), [-1,1].
+   * @param rot           Desired angular rate of the robot, [-1,1].
+   * @param fieldRelative Whether the provided x and y speeds are relative to the
+   *                      field.
+   * @param povAngleDeg   Get the angle in degrees of the D-pad (clockwise, -1
+   *                      means POV not pressed).
+   */
+  public void rotateOrKeepHeading(
+      double x, double y, double rot, boolean fieldRelative, int povAngleDeg) {
+    if (povAngleDeg != -1) {
+      targetHeadingDegrees = convertCardinalDirections(povAngleDeg);
+      keepHeading(x, y, fieldRelative);
+    } else if (rot == 0) {
+      keepHeading(x, y, fieldRelative);
+    } else {
+      targetHeadingDegrees = getHeadingDegrees();
+      drive(x, y, rot, fieldRelative);
+    }
+  }
+
+  public void setForward() {
+    frontLeft.setDesiredState(new SwerveModuleState(0, Rotation2d.fromDegrees(0)));
+    frontRight.setDesiredState(new SwerveModuleState(0, Rotation2d.fromDegrees(0)));
+    rearLeft.setDesiredState(new SwerveModuleState(0, Rotation2d.fromDegrees(0)));
+    rearRight.setDesiredState(new SwerveModuleState(0, Rotation2d.fromDegrees(0)));
+  }
+
+  public Pigeon2 getGyro() {
+    return gyro;
+  }
+
+  /** adapted from last year's version, which was taken from Github */
+  public Command followTrajectoryCommand(PathPlannerPath path, boolean isFirstPath) {
+    return new SequentialCommandGroup(
+        new InstantCommand(
+            () -> {
+              // Reset odometry for the first path you run during auto
+              if (isFirstPath) {
+                this.resetOdometry(path.getPreviewStartingHolonomicPose());
+              }
+            }),
+
+        new FollowPathHolonomic( // old method from last year: PPSwerveControllerCommand
+            path,
+            this::getPose, // pose supplier
+            this::getChassisSpeeds, // robot relative chassis speeds supplier
```

Comment:
Should we be getting the actual module speeds and converting to a chassis speeds here? Instead of just using the last set point?

Reply:
@avakimc